soliket/ccl.py

```diff
@@ -109,58 +122,70 @@ def calculate(self, state, want_derived=True, **params_values_dict):
         distance = self.provider.get_comoving_radial_distance(self.z)
         hubble_z = self.provider.get_Hubble(self.z)
         H0 = hubble_z[0]
+        h = H0/100
         E_of_z = hubble_z / H0
+
+        Omega_c = self.provider.get_param('omch2') / h ** 2
+        Omega_b = self.provider.get_param('ombh2') / h ** 2
         # Array z is sorted in ascending order. CCL requires an ascending scale factor
         # as input
         # Flip the arrays to make them a function of the increasing scale factor.
         # If redshift sampling is changed, check that it is monotonically increasing
         distance = np.flip(distance)
         E_of_z = np.flip(E_of_z)
 
-        # Create a CCL cosmology object
-        import pyccl as ccl
-        h = H0 / 100.
-        Omega_c = self.provider.get_param('omch2') / h ** 2
-        Omega_b = self.provider.get_param('ombh2') / h ** 2
-
-        # Currently, CCL requires the (ill-defined) linear "matter" perturbation
-        # growth factor and rate. Because it's ill-defined, we can't get it from
-        # Boltzmann code in general; ultimately CCL should use more physical
-        # inputs for anything of use in general models.
-        # For now just compute from CCL itself to keep it happy:
-        cosmo = ccl.Cosmology(Omega_c=Omega_c, Omega_b=Omega_b, h=h,
-                              n_s=self.provider.get_param('ns'),
-                              A_s=self.provider.get_param('As'))
         # Array z is sorted in ascending order. CCL requires an ascending scale
         # factor as input
         a = 1. / (1 + self.z[::-1])
-        growth = ccl.background.growth_factor(cosmo, a)
-        fgrowth = ccl.background.growth_rate(cosmo, a)
-        # In order to use CCL with input arrays, the cosmology object needs
-        # to be reset. This should be improved...
-        cosmo = ccl.Cosmology(Omega_c=Omega_c, Omega_b=Omega_b, h=h,
-                              n_s=self.provider.get_param('ns'),
-                              A_s=self.provider.get_param('As'), **self.extra_args)
-        cosmo._set_background_from_arrays(a_array=a,
-                                          chi_array=distance,
-                                          hoh0_array=E_of_z,
-                                          growth_array=growth,
-                                          fgrowth_array=fgrowth)
+        #growth = ccl.background.growth_factor(cosmo, a)
+        #fgrowth = ccl.background.growth_rate(cosmo, a)
+
+        # Create a CCL cosmology object. Because we are giving it background 
+        # quantities, it should not depend on the cosmology parameters given
+        cosmo = self.ccl.CosmologyCalculator(
+            Omega_c=Omega_c, Omega_b=Omega_b, h=h, sigma8=0.8, n_s=0.96,
```

Comment:
I would feel a bit more comfortable with `sigma8` and `n_s` coming from the provider, unless @damonge tells me it's 100% OK not to do so, not only for this current use case, but for plausible future ones.

Reply:
If you're passing linear power spectra, sigma8 and n_s should be redundant. You're not doing that here, but see my comment below.